File: examples/server/00_setup/01_new_account/go/main.go
// Package main demonstrates creating a brand new Self account from scratch.
// See README.md for detailed explanations and concepts.
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/joinself/academy/examples/go/common"
	"github.com/joinself/self-go-sdk/account"
	"github.com/joinself/self-go-sdk/event"
)

func main() {
	fmt.Println("🆕 New Account Creation Example")
	fmt.Println("===============================")

	// Check if account already exists
	fmt.Println("🔍 Checking for existing account...")
	if accountExists() {
		fmt.Println("⚠️  Account already exists in ./storage/")
		fmt.Println("💡 Delete ./storage/ to create fresh account, or use '../02_existing_account'")
		return
	}

	// Create new account
	fmt.Println("🔧 Creating new Self account...")
	selfAccount := createNewAccount()
	defer selfAccount.Close()

	// Display account information
	fmt.Println("\n📋 Account Information:")
	displayAccountInfo(selfAccount)
	displayStorageInfo()

	fmt.Println("\n✅ New Self account ready!")
}

func accountExists() bool {
	_, err := os.Stat("./storage")
	return !os.IsNotExist(err)
}

func createNewAccount() *account.Account {
	return common.SetupAccount(common.AccountConfig{
		Callbacks: account.Callbacks{
			OnConnect: func(acc *account.Account) {
				fmt.Println("✅ Connected to Self network")
			},
			OnMessage: func(selfAccount *account.Account, msg *event.Message) {
				// Message handler (unused in this example)
			},
		},
	})
}

func displayAccountInfo(selfAccount *account.Account) {
	inboxAddress, err := selfAccount.InboxOpen()
	if err != nil {
		log.Printf("⚠️ Could not open inbox: %v", err)
		return
	}

	fmt.Printf("🆔 Account DID: %s\n", inboxAddress.String())
	fmt.Printf("🔐 Status: Ready for secure communication\n")

	fmt.Println("\n💡 Your DID can be shared with others for connections")
}

func displayStorageInfo() {
	fmt.Println("\n📁 Storage: ./storage/ (encrypted)")
	fmt.Println("🔄 Use '../02_existing_account' to reload this account")
}

File: examples/server/00_setup/01_new_account/java/Main.kt
import com.joinself.selfsdk.kmp.account.Account
import com.joinself.selfsdk.kmp.account.LogLevel
import com.joinself.selfsdk.kmp.account.Target
import java.io.File
import java.util.concurrent.Semaphore

val signal = Semaphore(0)

fun main() {
    println("🆕 New Account Creation Example")
    println("===============================")

    // Check if account already exists
    println("🔍 Checking for existing account...")
    if (File("./storage").exists()) {
        println("⚠️  Account already exists in ./storage/")
        println("💡 Delete ./storage/ to create fresh account, or use '../02_existing_account'")
        return
    }

    // Create new account
    println("🔧 Creating new Self account...")
    val account = createAccount()
    
    // Display account information
    println("\n📋 New Account Information")
    println("==========================")
    val address = getAccountAddress(account)
    println("🆔 Account DID: $address")
    println("📬 Inbox Address: $address")
    println("🔐 Status: Encrypted and ready for secure communication")
    
    println("\n💡 Your DID can be shared with others for connections")
    
    println("\n📁 Storage: ./storage/ (encrypted)")
    println("🔄 Use '../02_existing_account' to reload this account")
    
    println("\n✅ New Self account ready!")
}

fun createAccount(): Account {
    val storageKey = "276cb6191a345753adb0897c2c0a89370aebf44ef99e612747bee3cd4e757ffa"
        .chunked(2).map { it.toInt(16).toByte() }.toByteArray()

    val account = Account()
    account.configure(
        storagePath = "./storage",
        storageKey = storageKey,
        rpcEndpoint = Target.PRODUCTION_SANDBOX.rpcEndpoint(),
        objectEndpoint = Target.PRODUCTION_SANDBOX.objectEndpoint(),
        messageEndpoint = Target.PRODUCTION_SANDBOX.messageEndpoint(),
        logLevel = LogLevel.WARN,
        onConnect = { 
            println("✅ Connected to Self network")
            println("✅ New account created successfully!")
            signal.release()
        },
        onDisconnect = { _ -> },
        onAcknowledgement = { _ -> },
        onError = { _, _ -> },
        onCommit = { _ -> },
        onKeyPackage = { _ -> },
        onWelcome = { _ -> },
        onProposal = { _ -> },
        onMessage = { _ -> },
        onIntegrity = null
    )

    signal.acquire() // Simple wait for connection
    return account
}

fun getAccountAddress(account: Account): String {
    var address = ""
    account.inboxOpen { status, addr -> 
        if (status.success()) address = addr.encodeHex()
        signal.release()
    }
    signal.acquire()
    return address
} 
